### DISCUSSION TITLE
7.0.0 causes a new PytestDeprecationWarning in my plugin, but I'm not sure why...

### DISCUSSION THREAD

**Discussion by wimglenn** [pytest-dev/pytest#9613] [Q&A] [answered]
I have a plugin, [pytest-raisin](https://github.com/wimglenn/pytest-raisin), which allows using exception instances with `pytest.raises`. For example:

```
import pytest

def test_example():
    with pytest.raises(KeyError("x")):
        {}["x"]
```

Since pytest 7.0.0, there's a new warning:

```
$ python -m pytest
============================================================= test session starts ==============================================================
platform darwin -- Python 3.10.1, pytest-7.0.0, pluggy-1.0.0
rootdir: /private/tmp/b
plugins: raisin-0.3
collected 1 item                                                                                                                               

test_example.py .                                                                                                                        [100%]

=============================================================== warnings summary ===============================================================
test_example.py::test_example
  /private/tmp/b/.venv/lib/python3.10/site-packages/pytest_raisin.py:91: PytestDeprecationWarning: A private pytest class or function was used.
    self.excinfo.__init__(tp)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================================= 1 passed, 1 warning in 0.00s =========================================================
```

The line it indicates is [here](https://github.com/wimglenn/pytest-raisin/blob/v0.3/pytest_raisin.py#L91). I took care not to use anything with a leading underscore in the plugin implementation. I've been using this plugin all over the place and it's very handy I think, it would be a shame if it got broken in pytest 8. Actually I still have some hopes that the idea might get accepted into pytest itself (originally proposed in https://github.com/pytest-dev/pytest/issues/4538).

Anything we could do either in pytest or in pytest-raisin to not trigger this warning? I had a look and https://github.com/pytest-dev/pytest/issues/7469 seems related, specifically the part that says:

> ~~_pytest.python_api.RaisesContext~~ - Not worth exposing, just a context manager.

**Comment by bluetech** [comment, marked as answer]
Hi @wimglenn,

You got it pretty much correctly - as part of formalizing pytest's public API, we audited `ExceptionInfo` and `RaisesContext`, and decided:

- Expose `ExceptionInfo`, but mark it's ctor private (so pytest is formally free to break how the type is constructed, but not any public methods/attributes).
- Don't expose `RaisesContext`.

My suggestion is to open an issue describing your use case for these, and we can make them public if it makes sense.

**Reply by wimglenn** [reply]
Thanks for the reply. Are all of ExceptionInfo's ctor private, or only the `__init__`? What about `ExceptionInfo.for_later()`?

**Reply by RonnyPfannschmidt** [reply]
ExceptionInfo is not a public api and I'd prefer if we had a opportunity to replace /remove it in future

**Reply by bluetech** [reply]
`ExceptionInfo.for_later` is public and [documented](https://docs.pytest.org/en/7.0.x/reference/reference.html#pytest.ExceptionInfo.for_later) which means that full stability is guaranteed for it, i.e. if we ever want to break it we will have to deprecate it first, and only remove in a major version.

**Reply by wimglenn** [reply]
In https://github.com/wimglenn/pytest-raisin/pull/7/files I've used `ExceptionInfo.for_later()` and the `fill_unfilled` method, hope that's OK.  It made the warnings go away..

**Reply by bluetech** [reply]
Now you're only using public API so should be good!
